File: backend/app/schemas/maintenance.py
"""Schemas for the library Maintenance feature (issue #147).

Covers the missing-cover table, the online cover-art search, and the beets
``unimported`` (stray file) detection + cleanup actions. These mirror the
snake_case album-family schemas (see ``schemas/album.py``) so the frontend
reads the same wire shape as the rest of the album endpoints.
"""
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class MissingCoverAlbum(BaseModel):
    album_id: int
    title: str
    artist: str
    # True when the album folder itself is gone from disk (ghost entry) —
    # cover search is pointless; the UI offers a DB-only removal instead.
    folder_missing: bool = False


class MissingCoverResponse(BaseModel):
    items: list[MissingCoverAlbum]
    total: int


class CoverSearchResult(BaseModel):
    url: str
    source: str
    width: Optional[int] = None
    height: Optional[int] = None


class CoverSearchResponse(BaseModel):
    query: str
    results: list[CoverSearchResult]


class PluginStatus(BaseModel):
    plugin: str
    enabled: bool


class StrayFile(BaseModel):
    path: str
    name: str
    size: int
    # True for image files — the UI offers a preview and use-as-cover.
    is_image: bool = False


class StrayGroup(BaseModel):
    folder: str
    relative_folder: str
    files: list[StrayFile]
    total_size: int
    # True when no file in this folder is tracked by beets — i.e. the whole
    # folder is a stray album, safe to move/delete wholesale.
    fully_untracked: bool
    # The single album living in this folder (None when untracked or when the
    # folder mixes several albums). Lets the UI show the album's current cover
    # next to stray images and offer use-as-cover.
    album_id: Optional[int] = None
    # mtime of that album's active cover, as a cache-buster for the cover URL.
    # None when the album has no cover yet.
    cover_version: Optional[int] = None


class UnimportedResponse(BaseModel):
    enabled: bool
    groups: list[StrayGroup]
    total_files: int


class StrayAction(str, Enum):
    delete = "delete"
    move_to_import = "move_to_import"


class StrayActionRequest(BaseModel):
    paths: list[str]
    action: StrayAction


class StrayActionResult(BaseModel):
    path: str
    status: str  # "deleted" | "moved" | "skipped" | "error"
    detail: Optional[str] = None
    relocated_to: Optional[str] = None


class StrayActionResponse(BaseModel):
    results: list[StrayActionResult]


class UseAsCoverRequest(BaseModel):
    """Promote a stray image file to the album cover of its folder."""

    path: str


class UseAsCoverResponse(BaseModel):
    status: str  # "cover_set"
    album_id: int
    cover_path: str


# --- BPM backfill (autobpm) ---


class BpmBackfillInfoResponse(BaseModel):
    """Pre-flight info for the maintenance page."""

    missing_count: int = Field(..., description="Tracks without a usable bpm tag")
    estimated_seconds: int = Field(
        0, description="Expected wall-clock duration of a full backfill"
    )
    workers: int = Field(1, description="Parallel analysis subprocesses that would be used")


class BpmBackfillStartResponse(BaseModel):
    job_id: str
    status: str = "queued"
    total: int


class BpmBackfillStatusResponse(BaseModel):
    """Current/last backfill job state for a library.

    ``status`` is one of: idle, queued, running, completed,
    completed_with_errors, cancelled, failed.
    """

    status: str
    total: int = 0
    processed: int = 0
    failed: int = 0
    job_id: Optional[str] = None
    error: Optional[str] = None
    updated_at: Optional[str] = None
    eta_seconds: Optional[int] = Field(
        None, description="Estimated seconds until the job finishes (measured rate)"
    )
    workers: Optional[int] = Field(None, description="Parallel analysis subprocesses in use")

File: frontend/src/api/maintenance.ts
// API client for the library Maintenance feature (issue #147).
// Mirrors the snake_case wire shape used by the album endpoints.

const API_URL = import.meta.env.VITE_API_URL || '/api'

export class MaintenanceApiError extends Error {
  status: number
  constructor(message: string, status: number) {
    super(message)
    this.name = 'MaintenanceApiError'
    this.status = status
  }
}

async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    let detail = 'An unknown error occurred'
    try {
      const data = await response.json()
      detail = data.detail ?? detail
    } catch {
      // body may not be JSON
    }
    throw new MaintenanceApiError(detail, response.status)
  }
  return response.json()
}

// ---------------------------------------------------------------------------
// Missing cover art
// ---------------------------------------------------------------------------
export interface MissingCoverAlbum {
  album_id: number
  title: string
  artist: string
  /** True when the album folder itself is gone from disk (ghost entry) —
   * cover search is pointless; the UI offers a DB-only removal instead. */
  folder_missing: boolean
}

export interface MissingCoverResponse {
  items: MissingCoverAlbum[]
  total: number
}

export async function fetchMissingCover(
  slug: string
): Promise<MissingCoverResponse> {
  const response = await fetch(
    `${API_URL}/v1/libraries/${encodeURIComponent(slug)}/maintenance/missing-cover`
  )
  return handleResponse<MissingCoverResponse>(response)
}

// ---------------------------------------------------------------------------
// Unimported (stray) files
// ---------------------------------------------------------------------------
export interface StrayFile {
  path: string
  name: string
  size: number
  /** True for image files — preview + use-as-cover are offered. */
  is_image: boolean
}

export interface StrayGroup {
  folder: string
  relative_folder: string
  files: StrayFile[]
  total_size: number
  fully_untracked: boolean
  /** The single album living in this folder, if it is a tracked album dir. */
  album_id: number | null
  /** mtime of that album's active cover (cache-buster); null = no cover. */
  cover_version: number | null
}

export interface UnimportedResponse {
  enabled: boolean
  groups: StrayGroup[]
  total_files: number
}

export type StrayAction = 'delete' | 'move_to_import'

export interface StrayActionResult {
  path: string
  status: string
  detail: string | null
  relocated_to: string | null
}

export interface StrayActionResponse {
  results: StrayActionResult[]
}

export interface PluginStatus {
  plugin: string
  enabled: boolean
}

export async function fetchUnimported(
  slug: string
): Promise<UnimportedResponse> {
  const response = await fetch(
    `${API_URL}/v1/libraries/${encodeURIComponent(slug)}/maintenance/unimported`
  )
  return handleResponse<UnimportedResponse>(response)
}

export async function enablePlugin(
  slug: string,
  plugin: string
): Promise<PluginStatus> {
  const response = await fetch(
    `${API_URL}/v1/libraries/${encodeURIComponent(slug)}/maintenance/plugins/${encodeURIComponent(plugin)}/enable`,
    { method: 'POST' }
  )
  return handleResponse<PluginStatus>(response)
}

export async function actOnStrays(
  slug: string,
  paths: string[],
  action: StrayAction
): Promise<StrayActionResponse> {
  const response = await fetch(
    `${API_URL}/v1/libraries/${encodeURIComponent(slug)}/maintenance/unimported/action`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ paths, action }),
    }
  )
  return handleResponse<StrayActionResponse>(response)
}

/** URL serving a stray image file for preview (no-store on the server). */
export function getStrayPreviewUrl(slug: string, path: string): string {
  return `${API_URL}/v1/libraries/${encodeURIComponent(slug)}/maintenance/unimported/preview?path=${encodeURIComponent(path)}`
}

export interface UseAsCoverResponse {
  status: string
  album_id: number
  cover_path: string
}

export async function useStrayAsCover(
  slug: string,
  path: string
): Promise<UseAsCoverResponse> {
  const response = await fetch(
    `${API_URL}/v1/libraries/${encodeURIComponent(slug)}/maintenance/unimported/use-as-cover`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ path }),
    }
  )
  return handleResponse<UseAsCoverResponse>(response)
}

// ---------------------------------------------------------------------------
// BPM backfill (autobpm)
// ---------------------------------------------------------------------------
export interface BpmInfo {
  missing_count: number
  estimated_seconds: number
  workers: number
}

export type BpmBackfillState =
  | 'idle'
  | 'queued'
  | 'running'
  | 'completed'
  | 'completed_with_errors'
  | 'cancelled'
  | 'failed'
  | 'unknown'

export interface BpmBackfillStatus {
  status: BpmBackfillState
  total: number
  processed: number
  failed: number
  job_id?: string | null
  error?: string | null
  updated_at?: string | null
  eta_seconds?: number | null
  workers?: number | null
}

export async function fetchBpmInfo(slug: string): Promise<BpmInfo> {
  const response = await fetch(
    `${API_URL}/v1/libraries/${encodeURIComponent(slug)}/maintenance/bpm`
  )
  return handleResponse<BpmInfo>(response)
}

export async function startBpmBackfill(slug: string): Promise<{ job_id: string; total: number }> {
  const response = await fetch(
    `${API_URL}/v1/libraries/${encodeURIComponent(slug)}/maintenance/bpm/backfill`,
    { method: 'POST' }
  )
  return handleResponse(response)
}

export async function fetchBpmBackfillStatus(slug: string): Promise<BpmBackfillStatus> {
  const response = await fetch(
    `${API_URL}/v1/libraries/${encodeURIComponent(slug)}/maintenance/bpm/backfill/status`
  )
  return handleResponse<BpmBackfillStatus>(response)
}

export async function cancelBpmBackfill(slug: string): Promise<void> {
  const response = await fetch(
    `${API_URL}/v1/libraries/${encodeURIComponent(slug)}/maintenance/bpm/backfill/cancel`,
    { method: 'POST' }
  )
  await handleResponse(response)
}
